### PR TITLE
TT-4856, added circuit breaker support for OAS API specification

### PR DESCRIPTION
**Pull request by andrei-tyk**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

**Comment by buger** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5899/merge
Commit:  
Triggered by: pull_request (@andrei-tyk)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7276477092)

**Comment by buger** [comment]
**API tests result - postgres15-sha256 env: success** :white_check_mark: 
Branch used: refs/pull/5899/merge
Commit:  
Triggered by: pull_request (@andrei-tyk)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7276477187)

**Comment by buger** [comment]
**API tests result - mongo44-sha256 env: success** :white_check_mark: 
Branch used: refs/pull/5899/merge
Commit:  
Triggered by: pull_request (@andrei-tyk)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7276477187)

**Comment by buger** [comment]
**API tests result - mongo44-murmur64 env: success** :white_check_mark: 
Branch used: refs/pull/5899/merge
Commit:  
Triggered by: pull_request (@andrei-tyk)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/7276477187)

File: apidef/oas/middleware.go [hunk 1]
```diff
@@ -1274,3 +1274,32 @@ func (e EndpointPostPlugins) ExtractTo(meta *apidef.GoPluginMeta) {
 	meta.PluginPath = e[0].Path
 	meta.SymbolName = e[0].Name
 }
+
+type CircuitBreaker struct {
```
Comment (jeffy-mathew) [inline apidef/oas/middleware.go:1278]: shall we add the corresponding Tyk classic api definition configuration in godoc sytle like https://github.com/TykTechnologies/tyk/blob/354bb28accd301f9d05d9fb36ca5064dc52e338b/apidef/oas/server.go#L14

File: apidef/oas/schema/x-tyk-api-gateway.json [hunk 2]
```diff
@@ -1415,6 +1415,38 @@
...
+      "type": "object",
+      "properties": {
+        "enabled": {
+          "type": "boolean"
+        },
+        "threshold": {
+          "type": "number",
+          "minimum": 0,
+          "maximum": 1,
+          "multipleOf": 0.01
```
Comment (jeffy-mathew) [inline apidef/oas/schema/x-tyk-api-gateway.json:1429]: 🥇

File: apidef/oas/middleware.go (hunk 1 above)
Comment (andrei-tyk) [reply apidef/oas/middleware.go:1278]: Added requested changes.

File: apidef/oas/middleware.go [hunk 3]
```diff
@@ -1274,3 +1274,37 @@ func (e EndpointPostPlugins) ExtractTo(meta *apidef.GoPluginMeta) {
...
```
Comment (titpetric) [inline apidef/oas/middleware.go:1278]: @andrei-tyk sorry, unresolving this (sadly on the same line)- struct missing godoc comment,

Maybe we can just link tyk classic apidefs on the CircuitBreaker struct godoc comment and not individual fields, if there's consensus with @jeffy-mathew (per-field migration docs seem excessive).

File: apidef/oas/middleware.go [hunk 4]
```diff
@@ -1274,3 +1274,37 @@ func (e EndpointPostPlugins) ExtractTo(meta *apidef.GoPluginMeta) {
...
+	// Enabled enables the Circuit Breaker functionality.
+	// Tyk classic API definition: `version_data.versions..extended_paths.circuit_breakers[*].disabled`.
+	Enabled bool `bson:"enabled" json:"enabled"`
+	// Threshold represents the percentage of requests that can error before the breaker is tripped. This must be a value between 0.0 and 1.0.
```
Comment (andyo-tyk) [inline apidef/oas/middleware.go:1282]:
```suggestion
	// Threshold is the proportion from each `sampleSize` requests that must fail for the breaker to be tripped. This must be a value between 0.0 and 1.0. If `sampleSize` is 100 then a threshold of 0.4 means that the breaker will be tripped if 40 out of every 100 requests fails.
```

File: apidef/oas/middleware.go [hunk 5]
```diff
@@ -1274,3 +1274,37 @@ func (e EndpointPostPlugins) ExtractTo(meta *apidef.GoPluginMeta) {
...
+	// Tyk classic API definition: `version_data.versions..extended_paths.circuit_breakers[*].threshold_percent`.
+	Threshold float64 `bson:"threshold" json:"threshold"`
+	// SampleSize represents the number of samples to take for a circuit breaker window.
```
Comment (andyo-tyk) [inline apidef/oas/middleware.go:1285]:
```suggestion
	// SampleSize is the size of the circuit breaker sampling window. Combining this with `threshold` gives the failure rate required to trip the circuit breaker.
```

File: apidef/oas/middleware.go [hunk 6]
```diff
@@ -1274,3 +1274,37 @@ func (e EndpointPostPlugins) ExtractTo(meta *apidef.GoPluginMeta) {
...
+	// Tyk classic API definition: `version_data.versions..extended_paths.circuit_breakers[*].samples`.
+	SampleSize int `bson:"sampleSize" json:"sampleSize"`
+	// CoolDownPeriod represents the amount of time (in seconds) needed to pass before returning to service.
```
Comment (andyo-tyk) [inline apidef/oas/middleware.go:1288]:
```suggestion
	// CoolDownPeriod is the period of time (in seconds) for which the circuit breaker will remain open before returning to service.
```

File: apidef/oas/middleware.go [hunk 7]
```diff
@@ -1274,3 +1274,37 @@ func (e EndpointPostPlugins) ExtractTo(meta *apidef.GoPluginMeta) {
...
+	// Tyk classic API definition: `version_data.versions..extended_paths.circuit_breakers[*].return_to_service_after`.
+	CoolDownPeriod int `bson:"coolDownPeriod" json:"coolDownPeriod"`
+	// HalfOpenStateEnabled allows some requests to pass through the circuit breaker during the cool down period.
```
Comment (andyo-tyk) [inline apidef/oas/middleware.go:1291]:
```suggestion
	// HalfOpenStateEnabled, if enabled, allows some requests to pass through the circuit breaker during the cool down period. If Tyk detects that the path is now working, the circuit breaker will be automatically reset and traffic will be resumed to the upstream.
```

**Review by andyo-tyk — CHANGES_REQUESTED** [review]
Proposed improvements to comments to improve clarity

**Review by titpetric — APPROVED** [review]

**Review by andyo-tyk — APPROVED** [review]
Great, thank you.

**Pull request merged**